pkg/wal/wal.go

```diff
@@ -42,19 +72,642 @@ type Segment interface {
 type WAL interface {
 	// Write a logging entity.
 	// It will return immediately when the data is written in the buffer,
-	// The returned function will be called when the entity is flushed on the persistent storage.
-	Write(seriesID common.SeriesID, timestamp time.Time, data []byte) (func(), error)
+	// The callback function will be called when the entity is flushed on the persistent storage.
+	Write(seriesID common.SeriesIDV2, timestamp time.Time, data []byte, callback func(common.SeriesIDV2, time.Time, []byte, error))
 	// Read specified segment by SegmentID.
-	Read(segmentID SegmentID) (*Segment, error)
+	Read(segmentID SegmentID) (Segment, error)
 	// ReadAllSegments reads all segments sorted by their creation time in ascending order.
-	ReadAllSegments() ([]*Segment, error)
+	ReadAllSegments() ([]Segment, error)
 	// Rotate closes the open segment and opens a new one, returning the closed segment details.
-	Rotate() (*Segment, error)
+	Rotate() (Segment, error)
 	// Delete the specified segment.
 	Delete(segmentID SegmentID) error
+	// Close all of segments and stop WAL work.
+	Close() error
+}
+
+// SegmentID identities a segment in a WAL.
+type SegmentID uint64
+
+// Segment allows reading underlying segments that hold WAl entities.
+type Segment interface {
+	GetSegmentID() SegmentID
+	GetLogEntries() []LogEntry
+}
+
+// LogEntry used for attain detail value of WAL entry.
+type LogEntry interface {
+	GetSeriesID() common.SeriesIDV2
+	GetTimestamps() []time.Time
+	GetValues() *list.List
+}
+
+// Log implements the WAL interface.
+type Log struct {
+	buffer           buffer
+	logger           *logger.Logger
+	bytesBuffer      *bytes.Buffer
+	timestampsBuffer *bytes.Buffer
+	segmentIndexMap  map[SegmentID]*segment
+	workSegment      *segment
+	writeChannel     chan logRequest
+	flushChannel     chan buffer
+	path             string
+	options          Options
+}
+
+type segment struct {
+	file       *os.File
+	path       string
+	logEntries []LogEntry
+	segmentID  SegmentID
+}
+
+type logRequest struct {
+	seriesID  common.SeriesIDV2
+	timestamp time.Time
+	callback  func(common.SeriesIDV2, time.Time, []byte, error)
+	data      []byte
+}
+
+type logEntry struct {
+	timestamps  []time.Time
+	values      *list.List
+	seriesID    common.SeriesIDV2
+	entryLength int64
+	count       int32
+}
+
+type buffer struct {
+	timestampMap map[common.SeriesIDV2][]time.Time
+	valueMap     map[common.SeriesIDV2][]byte
+	callbackMap  map[common.SeriesIDV2][]func(common.SeriesIDV2, time.Time, []byte, error)
+	count        int
 }
 
 // New creates a WAL instance in the specified path.
-func New(_ string, _ Options) (WAL, error) {
-	return nil, nil
+func New(path string, options *Options) (WAL, error) {
+	//  Check configuration options.
+	if options == nil {
+		options = DefaultOptions
+	}
+	if options.FileSize <= 0 {
+		options.FileSize = DefaultOptions.FileSize
```

Comment:
Instead of modifying the parameter, kindly create a local `options`.

Reply:
fixed